### PR TITLE
Add `WasmDocument::signDocument`

### PR DESCRIPTION
**Pull request by cycraig**
# Description of change
This adds the `signDocument` function to `WasmDocument` to allow single-update signing key rotations/replacements in the Wasm bindings. 

This is necessary because of breaking changes to the metadata which moved the `proof` location from the root level of the document into the `meta` sub-field. This means that `WasmDocument::signData` now places the `proof` in the wrong place in the document structure, at the root level. This method is not present on `IotaDocument` because its `sign_data` function is context-aware, and relies on the `SetSignature` trait to place the `proof` correctly, whereas `WasmDocument::signData` operates on arbitrary `JsValue` objects.

This is similar to why `WasmAccount` has a `createSignedDocument` function.

### Example

```ts
  const network = Network.mainnet();
  const config = Config.fromNetwork(network);
  const client = Client.fromConfig(config);

  // Create a new identity.
  const key = new KeyPair(KeyType.Ed25519);
  const doc = new Document(key, network.name);
  doc.signSelf(key, doc.defaultSigningMethod().id);
  const receipt = await client.publishDocument(doc);

  // Rotate the signing key, removing the previous method in a single update.
  let doc2 = Document.fromJSON(doc.toJSON());
  doc2.removeMethod(doc2.defaultSigningMethod().id);
  const newKey = new KeyPair(KeyType.Ed25519);
  const newMethod = new VerificationMethod(doc2.id, newKey.type, newKey.public, "sign-1");
  doc2.insertMethod(newMethod, MethodScope.CapabilityInvocation());
  doc2.metadataPreviousMessageId = receipt.messageId;
  doc2.metadataUpdated = Timestamp.nowUTC();

  // Sign the update with the previous document.

  // CURRENT: inserts the proof at the wrong location---root level alongside 'doc' and 'meta'.
  const signed = doc.signData(doc2.toJSON(), {
      method: doc.defaultSigningMethod().id,
      private: key.private,
  }, SignatureOptions.default());
  // Throws an error.
  try {
      Document.verifyDocument(Document.fromJSON(signed), doc)
  } catch(err) {
      console.log(err);
  }

  // THIS PR: correctly signs the document.
  doc.signDocument(doc2, key, doc.defaultSigningMethod().id);
  Document.verifyDocument(doc2, doc);

  // Publish the update.
  const updateReceipt = await client.publishDocument(doc2);
```

CURRENT (`signData`):
```json
{
  "doc": { ... },
  "meta": { ... },
  "proof": {
    "type": "JcsEd25519Signature2020",
    "verificationMethod": "#sign-0",
    "signatureValue": "4TZdrW4YHoYuaW4MzkKEMr91yAZVA4YWxzoJDSQWzSEFCyPDUE1aVT2PrW8hw7w87kcPW9vVrfnK26LyDt8WWZA2"
  }
}
```

THIS PR (`signDocument`):
```json
{
  "doc": { ... },
  "meta": { 
    ...
    "proof": {
       "type": "JcsEd25519Signature2020",
       "verificationMethod": "#sign-0",
       "signatureValue": "4TZdrW4YHoYuaW4MzkKEMr91yAZVA4YWxzoJDSQWzSEFCyPDUE1aVT2PrW8hw7w87kcPW9vVrfnK26LyDt8WWZA2"
     }
  },
}
```

### (1) Move `proof`
We can alternatively perform a breaking change to move the `proof` back to the root level, which would avoid this issue entirely. 

I.e.
```json
{
  "doc": { ... },
  "meta": { ... },
  "proof": { ... },
}
```

### (2) Downcast in `signData`
Downcast the `JsValue` parameter in `signData` to `WasmDocument`. However, downcasting requires unsafe inference via the constructor name right now, the signing would not be performed in-place, and the return type would still be `any` (unless we figure out convenient duck-typing somehow).

## Type of change

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] Enhancement (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Fix

## How the change has been tested
Manually tested with the example code and added a unit test.

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by cycraig** [comment]
Aside: should we make `verifyDocument` an associated method rather than a static function? I can't recall my reasoning for making it static, honestly, and the positional arguments are a bit confusing.

**Review by PhilippGackstatter — APPROVED** [review]
Looks good to me.

From the looks of it, `verify_document` would indeed benefit from being an associated method to make the distinction you mentioned clearer. I'll leave it up to you whether you want to do it in this PR or another one.

File: bindings/wasm/src/did/wasm_document.rs [hunk 1]
```diff
@@ -168,6 +169,35 @@ impl WasmDocument {
...
+  /// Signs another DID document using the verification method specified by `method_query`.
+  /// The `method_query` may be the full `DIDUrl` of the method or just its fragment,
+  /// e.g. "#sign-0".
+  ///
+  /// `Document.signSelf` should be used in general, this throws an error if trying to operate
+  /// on the same document. This is intended for signing updates to a document where a sole
+  /// capability invocation method is rotated or replaced entirely.
+  ///
+  /// NOTE: does not validate whether the private key of the given `key_pair` corresponds to the
+  /// verification method. See `Document::verifyDocument`.
```
Comment (abdulmth) [inline bindings/wasm/src/did/wasm_document.rs:181]:
```suggestion
  /// verification method. See {@link Document.verifyDocument}.
```

**Review by abdulmth — APPROVED** [review]
Looks good to me!

File: bindings/wasm/src/did/wasm_document.rs (hunk 1 above)
Comment (cycraig) [reply bindings/wasm/src/did/wasm_document.rs:181]: Yeah we should do a doc-comment pass in the Wasm bindings before 0.5. Does `@link` play nicely with markdown when rendered on the wiki?

Comment (cycraig) [reply bindings/wasm/src/did/wasm_document.rs:181]: Looks like it's expanded to a relative link in Markdown, which I think works.

`[verifyDocument](#Document.verifyDocument)`

**Pull request merged**